### PR TITLE
improve .withArgs error messages

### LINKED ISSUE #3439: Improve message for mismatched args in the `withArgs` matcher
If you do something like:

```js
await expect(contract.f()).to.emit(contract, "MyEvent").withArgs(2);
```

You get this error message:

```
      AssertionError: expected 1 to equal 2. The numerical values of the given "ethers.BigNumber" and "number" inputs were compared, and they differed.
      + expected - actual

      -1
      +2
```

But the message doesn't say anything about the assertion. It should include the fact that it's talking about an event, and include the name of the event. Something like:

```
      AssertionError: expected 1 to equal 2 in the first argument of the 'MyEvent' event. The numerical values of the given "ethers.BigNumber" and "number" inputs were compared, and they differed.
      + expected - actual

      -1
      +2
```

I'm also not convinced about the mentions of the types; that seems like an implementation detail we shouldn't show.

The same thing happens with the `revertedWithCustomError` matcher, and we should fix that too.

cc @shark0der

### PR DESCRIPTION
**Pull request by RenanSouza2**
Fixes #3439

- [X] Because this PR includes a **new feature**, the change was previously discussed on an Issue or with someone from the team.

### CODE REVIEW

**Comment by RenanSouza2** [comment]
Hey @fvictorio,

I was looking into the messages of the .withArgs and saw that custom error withArgs and events withArgs are very similar but defined separetadely,

I was thinking about unifying both of them using the same underlying function, what's your opinion on that?

Other thing, In this PR I'va already changed just a bit those functions because previously it was mixing the predicate error with the assertion error when the predicate returns false

**Comment by RenanSouza2** [comment]
Hey @fvictorio , I made this PR to improve the error messages in .withArgs, 

I ended up refactoring a larger chunk of the code that I initially expected, 

Let me know your comments or if this is not the path you wanna folow

File: packages/hardhat-chai-matchers/test/reverted/revertedWithCustomError.ts [hunk 1]
```diff
@@ -16,7 +16,7 @@ import { anyUint, anyValue } from "../../src/withArgs";
 import { MatchersContract } from "../contracts";
 
 describe("INTEGRATION: Reverted with custom error", function () {
-  describe("with the in-process hardhat network", function () {
+  describe.skip("with the in-process hardhat network", function () {
```
[the remark this answers is not in the document]
Comment (RenanSouza2) [reply packages/hardhat-chai-matchers/test/reverted/revertedWithCustomError.ts:19]: FOrgot to remove this skip

**Review by RenanSouza2 — COMMENTED** [review]
grammar

**Review by fvictorio — APPROVED** [review]

**Comment by fvictorio** [comment]
Thanks!

**Pull request merged**